File: app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/WebSearchAdapter.java
package org.fossasia.susi.ai.chat.adapters.recycleradapters;

import android.content.Context;
import android.net.Uri;
import android.support.customtabs.CustomTabsIntent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.squareup.picasso.Picasso;

import org.fossasia.susi.ai.R;
import org.fossasia.susi.ai.chat.adapters.viewholders.SearchResultHolder;
import org.fossasia.susi.ai.data.model.WebSearchModel;

import java.util.List;

import timber.log.Timber;

/**
 * <h1>Adapter to display horizontal list of web search results.</h1>
 * <p>
 * Created by chiragw15 on 2/6/17.
 */
public class WebSearchAdapter extends RecyclerView.Adapter<SearchResultHolder> {

    private LayoutInflater inflater;
    private Context context;
    private List<WebSearchModel> searchResults;

    /**
     * Instantiates a new Web search adapter.
     *
     * @param context       the context
     * @param searchResults the search results
     */
    public WebSearchAdapter(Context context, List<WebSearchModel> searchResults) {
        this.context = context;
        this.searchResults = searchResults;
        inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public SearchResultHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new SearchResultHolder(inflater.inflate(R.layout.search_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull final SearchResultHolder holder, int position) {
        holder.descriptionTextView.setVisibility(View.GONE);
        holder.titleTextView.setVisibility(View.GONE);
        holder.previewImageView.setVisibility(View.GONE);
        WebSearchModel webSearch = searchResults.get(position);
        if (webSearch != null) {
            String title = webSearch.getHeadline();
            String text = webSearch.getBody();
            String iconUrl = webSearch.getImageURL();
            final String linkurl = webSearch.getUrl();

            if (text != null) {
                holder.descriptionTextView.setText(text);
                holder.descriptionTextView.setVisibility(View.VISIBLE);
            } else {
                holder.descriptionTextView.setVisibility(View.GONE);
            }

            if (title != null) {
                holder.titleTextView.setText(title);
                holder.titleTextView.setVisibility(View.VISIBLE);
            } else {
                holder.titleTextView.setVisibility(View.GONE);
            }

            if (iconUrl != null && !iconUrl.isEmpty()) {
                holder.previewImageView.setVisibility(View.VISIBLE);
                Timber.v(iconUrl);
                Picasso.with(context).load(iconUrl)
                        .into(holder.previewImageView, new com.squareup.picasso.Callback() {
                            @Override
                            public void onSuccess() {
                                Timber.d("image loaded successfully");
                            }

                            @Override
                            public void onError() {
                                holder.previewImageView.setVisibility(View.GONE);
                            }
                        });

            } else {
                holder.previewImageView.setVisibility(View.GONE);
            }

            /*
              redirects to the url of the preview link through chrome custom tabs
             */
            holder.previewLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (linkurl != null) {
                        Uri webpage = Uri.parse(linkurl);
                        CustomTabsIntent.Builder builder = new CustomTabsIntent.Builder();  //custom tabs intent builder
                        CustomTabsIntent customTabsIntent = builder.build();
                        customTabsIntent.launchUrl(context, webpage); //launching through custom tabs
                    }
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return searchResults == null ? 0 : searchResults.size();
    }
}

File: app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/WebSearchAdapter.kt
package org.fossasia.susi.ai.chat.adapters.recycleradapters

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.support.customtabs.CustomTabsIntent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import org.fossasia.susi.ai.R
import org.fossasia.susi.ai.chat.adapters.viewholders.SearchResultHolder
import org.fossasia.susi.ai.data.model.WebSearchModel
import timber.log.Timber

/**
 * <h1>Adapter to display horizontal list of web search results.</h1>\
 */
class WebSearchAdapter(
    private val context: Context,
    private val searchResults: List<WebSearchModel?>?
) :
    RecyclerView.Adapter<SearchResultHolder>() {

    private val inflater: LayoutInflater = LayoutInflater.from(context)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SearchResultHolder {
        return SearchResultHolder(inflater.inflate(R.layout.search_item, parent, false))
    }

    override fun onBindViewHolder(holder: SearchResultHolder, position: Int) {
        holder.descriptionTextView.visibility = View.GONE
        holder.titleTextView.visibility = View.GONE
        holder.previewImageView.visibility = View.GONE
        val webSearch = searchResults?.get(position)
        if (webSearch != null) {
            val title = webSearch.headline
            val text = webSearch.body
            val iconUrl = webSearch.imageURL
            val linkurl = webSearch.url

            if (text != null) {
                holder.descriptionTextView.text = text
                holder.descriptionTextView.visibility = View.VISIBLE
            } else {
                holder.descriptionTextView.visibility = View.GONE
            }

            if (title != null) {
                holder.titleTextView.text = title
                holder.titleTextView.visibility = View.VISIBLE
            } else {
                holder.titleTextView.visibility = View.GONE
            }

            if (iconUrl != null && !iconUrl.isEmpty()) {
                holder.previewImageView.visibility = View.VISIBLE
                Timber.v(iconUrl)
                Glide.with(context)
                        .load(iconUrl)
                        .listener(object : RequestListener<Drawable>{
                            override fun onLoadFailed(e: GlideException?, model: Any?, target: Target<Drawable>?, isFirstResource: Boolean): Boolean {
                                holder.previewImageView.visibility = View.GONE
                                return true
                            }

                            override fun onResourceReady(resource: Drawable?, model: Any?, target: Target<Drawable>?, dataSource: DataSource?, isFirstResource: Boolean): Boolean {
                                Timber.d("image loaded successfully")
                                return true
                            }
                        })
                        .into(holder.previewImageView)
            } else {
                holder.previewImageView.visibility = View.GONE
            }

            /*
              redirects to the url of the preview link through chrome custom tabs
             */
            holder.previewLayout.setOnClickListener {
                if (linkurl != null) {
                    val webpage = Uri.parse(linkurl)
                    val builder = CustomTabsIntent.Builder() // custom tabs intent builder
                    val customTabsIntent = builder.build()
                    customTabsIntent.launchUrl(context, webpage) // launching through custom tabs
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return searchResults?.size ?: 0
    }
}
